sql/distsql: joinreader implementation

RaduBerinde opened the pull request:
See individual commits for descriptions.
CC @paperstreet, @nvanbenschoten

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7072)

<!-- Reviewable:end -->

RaduBerinde commented:
Reviewable note: please ignore r1 and r2.

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> #### sql/distsql: joinreader implementation
> 
> See individual commits for descriptions.
> 
> CC @paperstreet, @nvanbenschoten
> 
> 
> 
</details>


---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-:-KJbaJWbPFbS8ZHUF-rn:1113899997)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Reviewable note: had to resolve a conflict, you can ignore r1-r5.

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Reviewable note: please ignore r1 and r2.
</details>


---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-:-KJbsyVgToXVhOPAXbRf:-897311470)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on sql/distsql/base.go:
```diff
@@ -71,12 +71,16 @@ type RowChannel struct {
 
-func (rc *RowChannel) init() {
-	rc.dataChan = make(chan StreamMsg, rowChannelBuf)
+func (rc *RowChannel) initWithBufSize(chanBuf int) {
```
nit: naming of this parameter is slightly confusing. Why not `chanBufSize` and `rowChannelBufSize` (below)?

RaduBerinde replied:
Done.

spencerkimball commented on sql/distsql/joinreader_test.go:
```diff
@@ -0,0 +1,132 @@
+	//
+	//  |     a    |     b    |         sum         |         s           |
+	//  |-----------------------------------------------------------------|
+	//  | rowId/10 | rowId/10 | rowId/10 + rowId%10 | IntToEnglish(rowId) |
```
Under `b`, should be `rowId%10`, not `rowId/10` again.

RaduBerinde replied:
Thanks! Fixed.

tschottdorf commented:
Reviewed 2 of 2 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 7 of 7 files at r5, 7 of 7 files at r6, 1 of 1 files at r7, 7 of 7 files at r8, 7 of 7 files at r9, 1 of 1 files at r10, 8 of 8 files at r11.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[sql/distsql/data.proto, line 88 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfS9WPybnOnF0ZbusU:-KJfS9WPybnOnF0ZbusV:-1900056583) ([raw file](https://github.com/cockroachdb/cockroach/blob/dde14f124bca6bc92f61452ef3a1d70fe49e3be3/sql/distsql/data.proto#L88)):*
> ```protobuf
>   // via $0, $1, etc. If a secondary index is used, the columns that are not
>   // available as part of the index cannot be referenced.
>   optional Expression filter = 5 [(gogoproto.nullable) = false];
> ```

What's up with the numbering here?

---

*[sql/distsql/data.proto, line 124 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfSCFXWllTnwdGouSP:-KJfSCFXWllTnwdGouSQ:18861378) ([raw file](https://github.com/cockroachdb/cockroach/blob/dde14f124bca6bc92f61452ef3a1d70fe49e3be3/sql/distsql/data.proto#L124)):*
> ```protobuf
> // to interleave rows from multiple input streams.
> message InputSyncSpec {
>   // TODO(radu): add information for ordered synchronizers
> ```

nit: s/$/./

---

*[sql/distsql/data.proto, line 156 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfSLMIxOGDPBD_5xlc:-KJfSLMIxOGDPBD_5xld:-1905300810) ([raw file](https://github.com/cockroachdb/cockroach/blob/dde14f124bca6bc92f61452ef3a1d70fe49e3be3/sql/distsql/data.proto#L156)):*
> ```protobuf
> 
> message ProcessorSpec {
>   // In most cases, there is one input.
> ```

Is this comment (and the corresponding one below) useful?

---

*[sql/distsql/joinreader.go, line 52 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfSb94TvXJx58Lwl5h:-KJfSb94TvXJx58Lwl5i:2050597646) ([raw file](https://github.com/cockroachdb/cockroach/blob/dde14f124bca6bc92f61452ef3a1d70fe49e3be3/sql/distsql/joinreader.go#L52)):*
> ```Go
> 	if spec.IndexIdx != 0 {
> 		// TODO(radu): for now we only support joining with the primary index
> 		panic("join with index not implemented")
> ```

You could return an error.

---

*[sql/distsql/joinreader.go, line 104 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfSzDWTg1jSnqVecox:-KJfSzDWTg1jSnqVecoy:1125035736) ([raw file](https://github.com/cockroachdb/cockroach/blob/dde14f124bca6bc92f61452ef3a1d70fe49e3be3/sql/distsql/joinreader.go#L104)):*
> ```Go
> }
> 
> // minLoop runs the mainLoop and returns any error.
> ```

s/min/main/

---

*[sql/distsql/joinreader.go, line 170 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfTden9Mx9Pb4k2brv:-KJfTden9Mx9Pb4k2brw:-123769103) ([raw file](https://github.com/cockroachdb/cockroach/blob/dde14f124bca6bc92f61452ef3a1d70fe49e3be3/sql/distsql/joinreader.go#L170)):*
> ```Go
> func (jr *joinReader) Run(wg *sync.WaitGroup) {
> 	if log.V(1) {
> 		log.Infof("JoinReader filter: %s\n", jr.filter.expr)
> ```

Looks more verbose than `V(1)`.

---

*[sql/distsql/tablereader.go, line 86 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfMtN7CbZ0SbLOR31X:-KJfMtN7CbZ0SbLOR31Y:-1072380596) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf944f2071dd87f38f924c85aac659a5b515e3c8/sql/distsql/tablereader.go#L86)):*
> ```Go
> 	}
> 
> 	if indexIdx < 0 || indexIdx > len(rb.desc.Indexes) {
> ```

not `>=`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> Reviewable note: had to resolve a conflict, you can ignore r1-r5.
</details>


---

Review status: all files reviewed at latest revision, 14 unresolved discussions, all commit checks successful.

---

*[sql/distsql/data.proto, line 82 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfm6q79s4XHnoASdBB:-KJfm6q8Vtag1LQ3BOKw:1568809553) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/data.proto#L82)):*
> ```protobuf
>   // If 0, we use the primary index; each row in the input stream has a value
>   // for each primary key.
>   // TODO(radu): figure out the correct semantics when joining with an index.
> ```

what do you mean by this TODO? how is joining with a secondary index different than joining with a primary one?

---

*[sql/distsql/expr.go, line 77 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfZAdFhAoNop9D6-wU:-KJfZAdFhAoNop9D6-wV:-1384697272) ([raw file](https://github.com/cockroachdb/cockroach/blob/7e0dff84052f453ad50bc76cc2a10782d47f4107/sql/distsql/expr.go#L77)):*
> ```Go
> }
> 
> type exprHelper struct {
> ```

mind adding a comment to this struct

---

*[sql/distsql/joinreader.go, line 30 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfnGeUKo0gfBAPtjB3:-KJfnGeUKo0gfBAPtjB4:-1118666024) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/joinreader.go#L30)):*
> ```Go
> )
> 
> const joinReaderBatchSize = 100
> ```

add a TODO to batch per node ?

---

*[sql/distsql/joinreader.go, line 34 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfn-e5BVcWG5vvX6zm:-KJfn-e5BVcWG5vvX6zn:280313871) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/joinreader.go#L34)):*
> ```Go
> type joinReader struct {
> 	// RowChannel implements the rowReceiver interface.
> 	RowChannel
> ```

comment that the index key is constructed by using all the incoming cols, in order ?

---

*[sql/distsql/joinreader.go, line 144 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfoILJCbXipMczVByp:-KJfoILJCbXipMczVByq:-1448485314) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/joinreader.go#L144)):*
> ```Go
> 		}
> 
> 		for {
> ```

do the fetching async, and do a couple of them at a time? Or add a TODO? :)

---

*[sql/distsql/tablereader.go, line 86 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfMtN7CbZ0SbLOR31X:-KJf_gENsvchbmAaAo8K:-265394071) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf944f2071dd87f38f924c85aac659a5b515e3c8/sql/distsql/tablereader.go#L86)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> not `>=`?
</details>
indeed, see how `index` is  retrieved below.
But probably deserves a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-:-KJfpdqCfyvR1l6FMB1d:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[sql/distsql/data.proto, line 156 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfSLMIxOGDPBD_5xlc:-KJgQYUf5PtLHvu6EBSr:-1938593905) ([raw file](https://github.com/cockroachdb/cockroach/blob/dde14f124bca6bc92f61452ef3a1d70fe49e3be3/sql/distsql/data.proto#L156)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Is this comment (and the corresponding one below) useful?
</details>
It's common that processors have multiple input/output *streams*, most of them only have one InputSync/OutputRouter. I think the comment may help if someone is confused and thinks there is an InputSync for every input stream.

---

*[sql/distsql/data.proto, line 82 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfm6q79s4XHnoASdBB:-KJgPQqD-Td6-0E75po-:-2060137030) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/data.proto#L82)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> what do you mean by this TODO? how is joining with a secondary index different than joining with a primary one?
</details>
With a secondary index it's not clear if we need to support reading a non-unique index using just the index columns (allowing potentially many output rows per input row) or if we always require the implicit columns as well.

---

*[sql/distsql/joinreader.go, line 34 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfn-e5BVcWG5vvX6zm:-KJgV6G8RWVQCtVuYa1X:1015295689) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/joinreader.go#L34)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> comment that the index key is constructed by using all the incoming cols, in order ?
</details>
here, in the struct?

---

*[sql/distsql/joinreader.go, line 144 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfoILJCbXipMczVByp:-KJgVQv7mpCyKuR4Mz6t:-47805762) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/joinreader.go#L144)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> do the fetching async, and do a couple of them at a time? Or add a TODO? :)
</details>
I think that's more of a TODO for the rowfetcher than for joinreader no? (it should apply to tablereader just the same when there are many spans)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 9 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[sql/distsql/joinreader.go, line 34 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfn-e5BVcWG5vvX6zm:-KJgY1-p4yhxvJeSldNX:-1418072127) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/joinreader.go#L34)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> here, in the struct?
</details>
well, it's about how the data coming on the RowChannel is used. But I dunno, maybe a comment on the `joinReader` struct as a whole. I guess there's also a TODO in the proto about maybe making the input schema more flexible...

---

*[sql/distsql/joinreader.go, line 144 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfoILJCbXipMczVByp:-KJgYibOGarsBnCdTPE7:1586305111) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/joinreader.go#L144)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary>
> I think that's more of a TODO for the rowfetcher than for joinreader no? (it should apply to tablereader just the same when there are many spans)
</details>
Well we're consuming all the results for one span before calling the fetcher for the next one.
Maybe it's also a TODO for the row fetcher to become threadsafe and return some sort of a handle for each scan operation.

---

*[sql/distsql/joinreader.go, line 37 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJg_88w2FpJN4VPiUQR:-KJg_88w2FpJN4VPiUQS:1688708794) ([raw file](https://github.com/cockroachdb/cockroach/blob/f1f1b8a5f8c8849bf7cc7a4b7743580debc0ac9b/sql/distsql/joinreader.go#L37)):*
> ```Go
> type joinReader struct {
> 	// RowChannel implements the rowReceiver interface.
> 	RowChannel
> ```

wholly cow, you can embed multiple structs?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 9 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[sql/distsql/joinreader.go, line 144 \[r11\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJfoILJCbXipMczVByp:-KJgiacDz_qtLO8hrXcw:1846791959) ([raw file](https://github.com/cockroachdb/cockroach/blob/a481c1752c9d6dea0061ff8f9a94a07678f20eae/sql/distsql/joinreader.go#L144)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> Well we're consuming all the results for one span before calling the fetcher for the next one.
> Maybe it's also a TODO for the row fetcher to become threadsafe and return some sort of a handle for each scan operation.
</details>
Ah, got it. Added TODO.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072)*
<!-- Sent from Reviewable.io -->

paperstreet commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> :lgtm:
</details>


---

Review status: 0 of 9 files reviewed at latest revision, 17 unresolved discussions, all commit checks successful.

---

*[sql/distsql/joinreader.go, line 91 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJkbkEQuwSFmVrDRxCQ:-KJkbkERPLVMmbvAswsr:-1663302346) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573cf66833d657e35979b1911a0455bdf02441f/sql/distsql/joinreader.go#L91)):*
> ```Go
> 	}
> 
> 	// We will append to the prefix, so preallocate a larger size.
> ```

Here to the end seems very reusable (and parallel to EncodeIndexKey). Maybe pull it out into a function or add a TODO to pull it out

---

*[sql/distsql/joinreader.go, line 150 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJkcHx1fEiniy6IAJzk:-KJkcHx1fEiniy6IAJzl:2060609255) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573cf66833d657e35979b1911a0455bdf02441f/sql/distsql/joinreader.go#L150)):*
> ```Go
> 
> 		// TODO(radu): we are consuming all results from a fetch before starting
> 		// the next batch. We could overlap with the next fetch.
> ```

nit: "We could overlap with the next fetch." sounded to me on first read like a description of a possible race, instead of a possible optimization. Maybe tweak the wording

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-:-KJkdKqJe2J8Zxv5MQ8j:1417088721)*
<!-- Sent from Reviewable.io -->

RaduBerinde committed:
sql/distsql: move basic row fetching code into readerBase

Moving code that will be shared between tableReader and joinReader into
readerBase.

RaduBerinde committed:
sql/distsql: separate expression handling

Separating the tablereader code around the filter expression into an exprHelper,
which will be used by many processors.

RaduBerinde committed:
sql/distsql: joinreader implementation

Initial implementation of joinReader. This is a processor which takes in values
for key columns, looks up the corresponding rows, applies a filter, and outputs
desired columns from those rows. See the dist SQL RFC for more info.

RaduBerinde commented:
Review status: 0 of 10 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[sql/distsql/joinreader.go, line 91 \[r15\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072#-KJkbkEQuwSFmVrDRxCQ:-KJkjFZsn4xQiqA6O2tk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/7573cf66833d657e35979b1911a0455bdf02441f/sql/distsql/joinreader.go#L91)):*
<details><summary><i>Previously, paperstreet (Daniel Harrison) wrote…</i></summary>
> Here to the end seems very reusable (and parallel to EncodeIndexKey). Maybe pull it out into a function or add a TODO to pull it out
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7072)*
<!-- Sent from Reviewable.io -->
